src/ScalabilityRegionSignal.cpp

```diff
@@ -0,0 +1,120 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include "ScalabilityRegionSignal.hpp"
+
+#include "Exception.hpp"
+#include "Helper.hpp"
+#include "geopm_debug.hpp"
+
+#include <math.h>
+#include <unistd.h>
+
+namespace geopm
+{
+    ScalabilityRegionSignal::ScalabilityRegionSignal(std::shared_ptr<Signal> scalability_sig, std::shared_ptr<Signal> time_sig,
+                                         double range_upper, double range_lower,
+                                         double sleep_time)
+        : m_scalability(scalability_sig)
+        , m_time(time_sig)
+        , m_range_upper(range_upper)
+        , m_range_lower(range_lower)
+        , m_sleep_time(sleep_time)
+        , m_is_batch_ready(false)
+    {
+        GEOPM_DEBUG_ASSERT(m_scalability && m_time,
+                           "Signal pointers for scalability and time cannot be null.");
+    }
+
+    void ScalabilityRegionSignal::setup_batch(void)
+    {
+        if (!m_is_batch_ready) {
+            m_scalability->setup_batch();
+            m_time->setup_batch();
+            m_is_batch_ready = true;
+        }
+    }
+
+    double ScalabilityRegionSignal::compute_region_time(double scalability,
+                               double curr_time,
+                               double prev_time,
+                               double upper,
+                               double lower) {
+        double result = 0;
+        if (scalability < upper
+            && scalability >= lower
+            && !isnan(scalability)) {
+            result = curr_time - prev_time;
+        }
+
+        return result;
+    }
+
+    double ScalabilityRegionSignal::sample(void)
+    {
+        if (!m_is_batch_ready) {
+            throw Exception("setup_batch() must be called before sample().",
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        double scalability = m_scalability->sample();
+        double curr_time = m_time->sample();
+
+        m_region_time += compute_region_time(scalability,
+                                             curr_time,
+                                             m_prev_time,
+                                             m_range_upper,
+                                             m_range_lower);
+        m_prev_time = curr_time;
+
+        return m_region_time;
+    }
+
+    double ScalabilityRegionSignal::read(void) const
```

Comment:
I'm not sure this interface is useful except for testing a single step of sampling.